File: android/src/main/kotlin/com/example/native_in_app_purchase/NativeInAppPurchasePlugin.kt
package com.example.native_in_app_purchase

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class NativeInAppPurchasePlugin :
    FlutterPlugin,
    MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler,
    ActivityAware,
    BillingManager.Listener {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var applicationContext: Context
    private val mainHandler = Handler(Looper.getMainLooper())

    private var activity: Activity? = null
    private var eventSink: EventChannel.EventSink? = null
    private var billingManager: BillingManager? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
        billingManager = BillingManager(applicationContext, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        billingManager?.dispose()
        billingManager = null
        eventSink = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val manager = billingManager
        if (manager == null) {
            result.error("not_ready", "Billing manager is not available.", null)
            return
        }

        when (call.method) {
            "initialize" -> manager.initialize { error ->
                if (error == null) {
                    result.success(null)
                } else {
                    result.error(error.code, error.message, error.details)
                }
            }

            "isAvailable" -> manager.isAvailable { available, error ->
                if (error == null) {
                    result.success(available)
                } else {
                    result.error(error.code, error.message, error.details)
                }
            }

            "getProducts" -> {
                val productIds = call.argument<List<String>>("productIds").orEmpty()
                manager.getProducts(productIds) { payload, error ->
                    if (error == null) {
                        result.success(payload)
                    } else {
                        result.error(error.code, error.message, error.details)
                    }
                }
            }

            "buyProduct" -> {
                val productId = call.argument<String>("productId")
                val isConsumable = call.argument<Boolean>("isConsumable") ?: false
                val autoConsume = call.argument<Boolean>("autoConsume") ?: true
                val applicationUserName = call.argument<String>("applicationUserName")
                if (productId.isNullOrBlank()) {
                    result.error("invalid_product_id", "A non-empty productId is required.", null)
                    return
                }

                val currentActivity = activity
                if (currentActivity == null) {
                    result.error("no_activity", "An attached Activity is required to start a purchase flow.", null)
                    return
                }

                manager.buyProduct(
                    activity = currentActivity,
                    productId = productId,
                    isConsumable = isConsumable,
                    autoConsume = autoConsume,
                    applicationUserName = applicationUserName,
                ) { launched, error ->
                    if (error == null) {
                        result.success(launched)
                    } else {
                        result.error(error.code, error.message, error.details)
                    }
                }
            }

            "completePurchase" -> {
                val purchaseToken = call.argument<String>("purchaseToken")
                val productId = call.argument<String>("productId")
                val isConsumable = call.argument<Boolean>("isConsumable") ?: false
                manager.completePurchase(
                    purchaseToken = purchaseToken,
                    productId = productId,
                    isConsumable = isConsumable,
                ) { error ->
                    if (error == null) {
                        result.success(null)
                    } else {
                        result.error(error.code, error.message, error.details)
                    }
                }
            }

            "restorePurchases" -> manager.restorePurchases { error ->
                if (error == null) {
                    result.success(null)
                } else {
                    result.error(error.code, error.message, error.details)
                }
            }

            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        billingManager?.setActivity(binding.activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
        billingManager?.setActivity(null)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        billingManager?.setActivity(binding.activity)
    }

    override fun onDetachedFromActivity() {
        activity = null
        billingManager?.setActivity(null)
    }

    override fun onPurchaseUpdate(purchase: Map<String, Any?>) {
        mainHandler.post {
            eventSink?.success(purchase)
        }
    }

    override fun onPurchaseError(error: BillingManager.PluginError) {
        mainHandler.post {
            eventSink?.error(error.code, error.message, error.details)
        }
    }

    companion object {
        private const val METHOD_CHANNEL = "native_in_app_purchase"
        private const val EVENT_CHANNEL = "native_in_app_purchase/events"
    }
}

File: ios/Classes/NativeInAppPurchasePlugin.swift
import Flutter
import UIKit

public class NativeInAppPurchasePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private let storeKitManager = StoreKitManager()
  private var eventSink: FlutterEventSink?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let methodChannel = FlutterMethodChannel(
      name: "native_in_app_purchase",
      binaryMessenger: registrar.messenger()
    )
    let eventChannel = FlutterEventChannel(
      name: "native_in_app_purchase/events",
      binaryMessenger: registrar.messenger()
    )

    let instance = NativeInAppPurchasePlugin()
    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    eventChannel.setStreamHandler(instance)
  }

  override init() {
    super.init()
    storeKitManager.delegate = self
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      storeKitManager.initialize()
      result(nil)
    case "isAvailable":
      result(storeKitManager.isAvailable())
    case "getProducts":
      guard
        let arguments = call.arguments as? [String: Any],
        let productIds = arguments["productIds"] as? [String]
      else {
        result(
          FlutterError(
            code: "invalid_arguments",
            message: "Expected a productIds array.",
            details: nil
          )
        )
        return
      }

      storeKitManager.getProducts(productIds: productIds, result: result)
    case "buyProduct":
      guard
        let arguments = call.arguments as? [String: Any],
        let productId = arguments["productId"] as? String
      else {
        result(
          FlutterError(
            code: "invalid_arguments",
            message: "Expected a productId value.",
            details: nil
          )
        )
        return
      }

      let applicationUserName = (call.arguments as? [String: Any])?["applicationUserName"] as? String
      let isConsumable = (call.arguments as? [String: Any])?["isConsumable"] as? Bool ?? false
      let autoConsume = (call.arguments as? [String: Any])?["autoConsume"] as? Bool ?? true

      storeKitManager.buyProduct(
        productId: productId,
        applicationUserName: applicationUserName,
        isConsumable: isConsumable,
        autoConsume: autoConsume,
        result: result
      )
    case "completePurchase":
      guard let arguments = call.arguments as? [String: Any] else {
        result(
          FlutterError(
            code: "invalid_arguments",
            message: "Expected completePurchase arguments.",
            details: nil
          )
        )
        return
      }

      storeKitManager.completePurchase(
        transactionId: arguments["transactionId"] as? String,
        result: result
      )
    case "restorePurchases":
      let arguments = call.arguments as? [String: Any]
      let consumableProductIds = arguments?["consumableProductIds"] as? [String] ?? []
      storeKitManager.restorePurchases(
        consumableProductIds: consumableProductIds,
        result: result
      )
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink)
    -> FlutterError?
  {
    eventSink = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }
}

extension NativeInAppPurchasePlugin: StoreKitManagerDelegate {
  func storeKitManager(_ manager: StoreKitManager, didUpdatePurchase purchase: [String: Any?]) {
    eventSink?(purchase)
  }

  func storeKitManager(_ manager: StoreKitManager, didFailWith error: FlutterError) {
    eventSink?(error)
  }
}
